text/3366-diagnostic-attribute-namespace.md

```diff
@@ -0,0 +1,222 @@
+
+This section describes the syntax of the `on_unimplemented` attribute and additionally how it is supposed to work. Implementing the later part is optional as outlined in the previous section. Any syntax not explicitly described in this section must be rejected as invalid according to the general rules of the `#[diagnostic]` attribute namespace.
+
+```rust
+#[diagnostic::on_unimplemented(
+    message="message",
+    label="label",
+    note="note",
+)]
+trait MyIterator<A> {
+    fn next(&mut self) -> A;
+}
+```
+
+
+Each of the options `message`, `label` and `note` are optional. They are separated by comma. The tailing comma is optional. Specifying any of these options hints the compiler to replace the normally emitted part of the error message with the provided string. At least one of these options needs to exist. Each option can appear at most once. The error message can include type information for the `Self` type or any generic type by using `{Self}` or `{A}` (where `A` refers to the generic type name in the definition). These placeholders are replaced by the actual type name.
+
+In addition the `on_unimplemented` attribute provides mechanisms to specify for which exact types a certain message should be emitted via an `if()` option. It accepts a set of filter options. A filter option consists on the generic parameter name from the trait definition and a type path against which the parameter should be checked. These type path could either be a fully qualified path or refer to any type in the current scope. As a special generic parameter name `Self` is added to refer to the `Self` type of the trait implementation. A filter option evaluates to `true` if the corresponding generic parameter in the trait definition matches the specified type. The provided `message`/`note`/`label` options are only emitted if the filter operation evaluates to `true`.
+
+The `any` and `all` option allows to combine multiple filter options. The `any` option matches if one of the supplied filter options evaluates to `true`, the `all` option requires that all supplied filter options evaluate to true. `not` allows to negate a given filter option. It evaluates to `true` if the inner filter option evaluates to `false`. These options can be nested to construct complex filters.
+
+The `on_unimplemented` attribute can be applied multiple times to the same trait definition. Multiple attributes are evaluated in order. The first matching instance for each of the `message`/`label`/`note` options is emitted.
+```rust
+#[diagnostic::on_unimplemented(
+    if(Self = std::string::String),
+    note = "That's only emitted if Self == std::string::String"
+)]
+#[diagnostic::on_unimplemented(
+    if(A = String), // Refers to whatever `String` is in the current scope
+    note = "That's only emitted if A == String",
+)]
+#[diagnostic::on_unimplemented(
+    if(any(A = i32, Self = i32)),
+    note = "That's emitted if A or Self is a i32",
+)]
```

Comment:
Should this check whether a type has the *name* String or whether it *is* String? `#[rustc_on_unimplemented]` [blindly checks the name of the type](https://play.integer32.com/?version=stable&mode=debug&edition=2021&gist=a24ff48dc2174de7692484b622c4f6c2), but "Refers to whatever `String` is in the current scope" sounds like you want the latter.

Reply:
I would anticipate that `on_unimplemented` would receive the `DefId` of whatever path was passed in here, so it would be following regular visibility rules.

Reply:
I propose to use the regular visibility rules here instead of just matching the name as string as that's currently done. I feel that aligns the proposed attribute more with other parts of rust.